pkg/sources/job_progress_hook.go

```diff
@@ -122,51 +175,38 @@ func (u *UnitHook) ReportError(ref JobProgressRef, err error) {
 }
 
 func (u *UnitHook) Finish(ref JobProgressRef) {
-	u.mu.Lock()
-	defer u.mu.Unlock()
 	// Clear out any metrics on this job. This covers the case for the
 	// source running without unit support.
-	prefix := u.id(ref, nil)
-	for _, id := range u.metrics.Keys() {
-		if !strings.HasPrefix(id, prefix) {
-			continue
-		}
-		metric, ok := u.metrics.Get(id)
-		if !ok {
-			continue
-		}
-		// If the unit is nil, the source does not support units.
-		// Use the overall job metrics instead.
-		if metric.Unit == nil {
-			snap := ref.Snapshot()
-			metric.StartTime = snap.StartTime
-			metric.EndTime = snap.EndTime
-			metric.Errors = snap.Errors
-		}
+	id := u.id(ref, nil)
+	metrics, ok := u.finishUnit(id)
+	if !ok {
+		return
 	}
+	snap := ref.Snapshot()
+	metrics.StartTime = snap.StartTime
+	metrics.EndTime = snap.EndTime
+	metrics.Errors = snap.Errors
+	u.ejectFinishedMetric(*metrics)
 }
 
-// UnitMetrics gets all the currently active or newly finished metrics for this
-// job. If a unit returned from this method has finished, it will be removed
-// from the cache and no longer returned in successive calls to UnitMetrics().
-func (u *UnitHook) UnitMetrics() []UnitMetrics {
+// InProgressSnapshot gets all the currently active metrics across all jobs.
+func (u *UnitHook) InProgressSnapshot() []UnitMetrics {
 	u.mu.Lock()
 	defer u.mu.Unlock()
 	output := make([]UnitMetrics, 0, u.metrics.Len())
 	for _, id := range u.metrics.Keys() {
-		metric, ok := u.metrics.Get(id)
-		if !ok {
-			continue
-		}
-		output = append(output, *metric)
-		if metric.IsFinished() {
-			metric.handled = true
-			u.metrics.Remove(id)
+		if metrics, ok := u.metrics.Get(id); ok {
+			output = append(output, *metrics)
 		}
 	}
 	return output
 }
 
+func (u *UnitHook) Close() error {
```

Comment:
Is this used anywhere?

Reply:
Yes, it gets called when the `SourceManager` is waited:

https://github.com/trufflesecurity/trufflehog/blob/fb230aa0ae379066cd6dd348d723a0eb7ba0fb47/pkg/sources/source_manager.go#L173-L177

Reply:
Some synchronization is required to make the goroutine reading the channel finish before everything gets shut down, but that doesn't happen in this PR since `UnitHook` isn't used yet.

I figured it made sense for the source manager to close out the hooks since the hook initialized it.

Reply:
> Some synchronization is required to make the goroutine reading the channel finish before everything gets shut down, but that doesn't happen in this PR since UnitHook isn't used yet.

You read my mind :)